mapping: add TLB for dosaddr_to_unixaddr for unprotected pages

bartoldeman opened the pull request:
commit 705dab48 made dosaddr_to_unixaddr much slower. It's on the hot path for cpusim (read_byte etc.), so this slowed it down by almost 50% for code running from low memory.

I could make dosaddr_to_unixaddr faster again using a small hack:
```
diff --git a/src/arch/linux/mapping/mapping.c b/src/arch/linux/mapping/mapping.c
index ed40a6b11..708384308 100644
--- a/src/arch/linux/mapping/mapping.c
+++ b/src/arch/linux/mapping/mapping.c
@@ -108,13 +108,11 @@ static void update_aliasmap(dosaddr_t dosaddr, size_t mapsize,
   hwram_update_aliasmap(hw, addr2, mapsize, unixaddr);
 }

+static unsigned char **lowmem_aliasmap;
 void *dosaddr_to_unixaddr(dosaddr_t addr)
 {
-  if (addr < ALIAS_SIZE) {
-    void *ret = get_hardware_uaddr(addr);
-    if (ret != MAP_FAILED)
-      return ret;
-  }
+  if (addr < ALIAS_SIZE && lowmem_aliasmap[addr >> PAGE_SHIFT])
+    return lowmem_aliasmap[addr >> PAGE_SHIFT] + (addr & (PAGE_SIZE - 1));
   return MEM_BASE32(addr);
 }

@@ -854,6 +852,8 @@ static int do_register_hwram(int type, unsigned base, unsigned size,
   hw->size = size;
   hw->type = type;
   hw->aliasmap = alloc_aliasmap(uaddr, size);
+  if (type == 'L')
+    lowmem_aliasmap = hw->aliasmap;
   hw->next = hardware_ram;
   hardware_ram = hw;
   if (!uaddr && (base >= LOWMEM_SIZE || type == 'h'))
```

but even faster is to use a TLB, since we already have the infrastructure and invalidation logic to see if the page is special, and profiling then shows dosaddr_to_unixaddr is only used very rarely.

Here's my little testcase (nasm -fbin test.asm -o test.com), just writing the same byte 268 million times:

```
org 100h
mov bx, 800h
outer:
mov cx, 0ffffh
inner:
mov byte [test], 0
loop inner
dec bx
jnz outer
ret
test:
```

Test with cpusim:
Before: 22.7s (11.8 MIPS)
With quick and dirty patch above: 16.0s (16.8 MIPS) With TLB: 14.5s (18.5 MIPS)

bartoldeman committed:
mapping: add TLB for dosaddr_to_unixaddr for unprotected pages

commit 705dab48 made dosaddr_to_unixaddr much slower. It's on the hot path
for cpusim (read_byte etc.), so this slowed it down by almost 50% for code
running from low memory.

I could make dosaddr_to_unixaddr faster again using a small hack:
```
diff --git a/src/arch/linux/mapping/mapping.c b/src/arch/linux/mapping/mapping.c
index ed40a6b11..708384308 100644
--- a/src/arch/linux/mapping/mapping.c
+++ b/src/arch/linux/mapping/mapping.c
@@ -108,13 +108,11 @@ static void update_aliasmap(dosaddr_t dosaddr, size_t mapsize,
   hwram_update_aliasmap(hw, addr2, mapsize, unixaddr);
 }

+static unsigned char **lowmem_aliasmap;
 void *dosaddr_to_unixaddr(dosaddr_t addr)
 {
-  if (addr < ALIAS_SIZE) {
-    void *ret = get_hardware_uaddr(addr);
-    if (ret != MAP_FAILED)
-      return ret;
-  }
+  if (addr < ALIAS_SIZE && lowmem_aliasmap[addr >> PAGE_SHIFT])
+    return lowmem_aliasmap[addr >> PAGE_SHIFT] + (addr & (PAGE_SIZE - 1));
   return MEM_BASE32(addr);
 }

@@ -854,6 +852,8 @@ static int do_register_hwram(int type, unsigned base, unsigned size,
   hw->size = size;
   hw->type = type;
   hw->aliasmap = alloc_aliasmap(uaddr, size);
+  if (type == 'L')
+    lowmem_aliasmap = hw->aliasmap;
   hw->next = hardware_ram;
   hardware_ram = hw;
   if (!uaddr && (base >= LOWMEM_SIZE || type == 'h'))
```

but even faster is to use a TLB, since we already have the infrastructure and
invalidation logic to see if the page is special, and profiling then shows
dosaddr_to_unixaddr is only used very rarely.

To make this easier to follow, mem_likely_protected() has been renamed to
unprotected_dosaddr_to_unixaddr()

Here's my little testcase, just writing the same byte 268 million times:

org 100h
mov bx, 800h
outer:
mov cx, 0ffffh
inner:
mov byte [test], 0
loop inner
dec bx
jnz outer
ret
test:

Test with cpusim:
Before: 22.7s (11.8 MIPS)
With quick and dirty patch above: 16.0s (16.8 MIPS)
With TLB: 14.5s (18.5 MIPS)

stsp commented on src/base/misc/dos2linux.c:
```diff
@@ -531,9 +533,17 @@ static inline int mem_likely_protected(dosaddr_t addr, int len)
 
+static inline void *unprotected_dosaddr_to_unixaddr(dosaddr_t addr)
+{
+  int hash = (addr >> PAGE_SHIFT) & (PAGE_SIZE-1);
```
Don't you need to check here that
hash not collided, by checking the
`unprotected_page_cache[hash] == addr & PAGE_MASK`
first?

bartoldeman replied:
this is already done by `mem_likely_protected` in the callers.
If I do it here as well, it would be a redundant check.

I think however I can have `mem_likely_protected` return a unix addr (should rename it too then) to make this more clear.

bartoldeman replied:
I missed a case in fact, as seen by CI in `test/test_dos.py PPDOSGITTestCase.test_memory_emm286_borland`. I've restructured it now so it's crystal clear that no colliding hashes are returned.

bartoldeman replied:
CI fails the cpu tests on MSDOS 6.22 now but it's ok with FDPP. A bit unusual but will check!

bartoldeman replied:
MSDOS was failing with the HMA; its mapping wasn't invalidated when it changed a20. This wasn't an issue before as it remained rwx but now of course the unix address changes.

it turned out the invalidation was done from the wrong place (my fault!). I changed it to call it from aliasing and mprotect functions in mapping.c now.

bartoldeman committed:
mapping: fix invalidate_page_cache() caller positions

The invalidation was done from do_invalidate in simx86, which isn't
correct since that only unprotects pages with JIT code.

Instead call it when the alias mapping or mprotect changes,
and (the other way around), call e_invalidate from the write
functions because they may write to JIT-protected pages.

stsp commented on src/base/misc/dos2linux.c:
```diff
@@ -528,12 +532,23 @@ static inline int mem_likely_protected(dosaddr_t addr, int len)
+  unprotected_page_cache[hash] = addr & PAGE_MASK;
+  uaddr = dosaddr_to_unixaddr(addr & PAGE_MASK);
+  unprotected_page_unixaddr_tlb[hash] = uaddr;
+  return &uaddr[addr & (PAGE_SIZE-1)];
```
That func is definitely not the best engineering.
Maybe you mixed so many things in 1 func for
a speed-up. But I think if you move
dosaddr_to_unixaddr() to the caller, and pass
uaddr to that func rather than return it, then
the speed would be the same or even better.

stsp replied:
So I suggest to keep having set_unprotected_page()
and just pass an uaddr to it as an extra arg.

bartoldeman replied:
Done

bartoldeman replied:
For the speed it doesn't really matter. This is all the slow path, the fast path only calls `unprotected_dosaddr_to_unixaddr()`.

stsp commented on src/base/misc/dos2linux.c:
```diff
@@ -556,28 +571,30 @@ static void check_read_pagefault(dosaddr_t addr,
     if (config.mmio_tracing && mmio_check(addr))
       return mmio_trace_byte(addr, READ_BYTE(addr), MMIO_READ);
-    check_read_pagefault(addr, handler);
+    uaddr = check_read_pagefault(addr, handler);
```
Would it be possible to split that
uaddr things into a separate patch?
I never looked into your caching code
before, so I can't grok large patches in
that area. :)

bartoldeman replied:
The present patch is a bit smaller, so is it ok as is. I'll have a look how to still split it up, but it's not obvious to me.

stsp commented on src/base/misc/dos2linux.c:
```diff
@@ -615,6 +615,7 @@ uint64_t do_read_qword(dosaddr_t addr, sim_pagefault_handler_t handler)
 static int check_write_pagefault(dosaddr_t addr, uint32_t op, int len,
 				 sim_pagefault_handler_t handler)
 {
+  e_invalidate(addr, len);
```
Maybe move that to the callers?
`check_write_pagefault()` just doesn't suggest
(by its name) to be the right place for things
that should work no matter whether the
write should page-faulted.

Also while looking at the callers, I've found
this code:
```
void do_write_dword(dosaddr_t addr, uint32_t dword, sim_pagefault_handler_t hand
ler)
{
  if (mem_likely_protected(addr, 4)) {
    if (((addr+3) & (PAGE_SIZE-1)) < 3) {
      do_write_word(addr, dword & 0xffff, handler);
      do_write_word(addr+2, dword >> 16, handler);
    }
```
Isn't a `return` missing after splitting the write?

bartoldeman replied:
all done now

bartoldeman committed:
mapping: add missing return statement for unaligned writes

stsp commented:
Thanks!
Will see if that speeds up an arm port.
